authorizer/annotation.go

```diff
@@ -0,0 +1,171 @@
+func (s *AnnotationService) DeleteAnnotations(ctx context.Context, orgID platform.ID, delete influxdb.AnnotationDeleteFilter) error {
+	if _, _, err := AuthorizeOrgWriteResource(ctx, influxdb.AnnotationsResourceType, orgID); err != nil {
+		return err
+	}
+	return s.s.DeleteAnnotations(ctx, orgID, delete)
+}
+
+// DeleteAnnotation checks to see if the authorizer on context has write access to the requested annotation
+func (s *AnnotationService) DeleteAnnotation(ctx context.Context, id platform.ID) error {
+	a, err := s.s.GetAnnotation(ctx, id)
+	if err != nil {
+		return err
+	}
+	if _, _, err := AuthorizeWrite(ctx, influxdb.AnnotationsResourceType, id, a.OrgID); err != nil {
+		return err
+	}
+	return s.s.DeleteAnnotation(ctx, id)
+}
+
+// UpdateAnnotation checks to see if the authorizer on context has write access to the requested annotation
+func (s *AnnotationService) UpdateAnnotation(ctx context.Context, id platform.ID, update influxdb.AnnotationCreate) (*influxdb.AnnotationEvent, error) {
+	a, err := s.s.GetAnnotation(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if _, _, err := AuthorizeWrite(ctx, influxdb.AnnotationsResourceType, id, a.OrgID); err != nil {
+		return nil, err
+	}
+	return s.s.UpdateAnnotation(ctx, id, update)
+}
+
+// ListStreams checks to see if the authorizer on context has read access for streams for the provided orgID
+func (s *AnnotationService) ListStreams(ctx context.Context, orgID platform.ID, filter influxdb.StreamListFilter) ([]influxdb.ReadStream, error) {
+	if _, _, err := AuthorizeOrgReadResource(ctx, influxdb.AnnotationsResourceType, orgID); err != nil {
+		return nil, err
+	}
+
+	return s.s.ListStreams(ctx, orgID, filter)
+}
+
+// GetStream checks to see if the authorizer on context has read access to the requested stream
+func (s *AnnotationService) GetStream(ctx context.Context, id platform.ID) (*influxdb.StoredStream, error) {
+	st, err := s.s.GetStream(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if _, _, err := AuthorizeRead(ctx, influxdb.AnnotationsResourceType, id, st.OrgID); err != nil {
+		return nil, err
+	}
+	return st, nil
+}
+
+func (s *AnnotationService) CreateOrUpdateStream(ctx context.Context, orgID platform.ID, stream influxdb.Stream) (*influxdb.ReadStream, error) {
+	// We need to know if the request is creating a new stream, or updating an existing stream to check
+	// permissions appropriately
+
+	// Get the stream by name. An empty slice will be returned if the stream doesn't exist
+	// note: a given org can only have one stream by the same name. this constraint is enforced in the database schema
+	streams, err := s.s.ListStreams(ctx, orgID, influxdb.StreamListFilter{
+		StreamIncludes: []string{stream.Name},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// for updating an existing stream, we need to get the stream ID before calling the UpdateStream method
+	if len(streams) != 0 {
```

Comment:
Could we instead do:
```suggestion
	if len(streams) == 1 {
```
Even though I read the comment above about only having 1 stream per org with a given name, my brain still reacted like "what about the rest of the `streams`?" when I saw `streams[o].ID` below.

Reply:
Seems like another reasonable edge case to guard against, so I updated the code to do specific things on `==1` or `==0`, and return a server error if more than one matching stream is returned somehow.